### PR TITLE
Implement Get and List and other improvements

### PR DESCRIPTION
**Pull request by cjellick**
This PR:
- Implements the new Get and List docker volume driver methods
- Refactors Longhorn significantly
- makes some other small changes/fixes

Addresses:
- https://github.com/rancher/rancher/issues/3568
- https://github.com/rancher/rancher/issues/3671

### CODE REVIEW

File: api/response.go [hunk 1]
```diff
@@ -57,12 +58,17 @@ func ResponseLogAndError(v interface{}) {
...
-		logrus.Errorf("%s: %s", v, debug.Stack())
-		ResponseError("Caught FATAL error: %s", v)
+		e, isErr := v.(error)
+		_, isRuntimeErr := e.(runtime.Error)
+		if isErr && !isRuntimeErr {
+			logrus.Errorf(fmt.Sprint(e))
+			ResponseError(fmt.Sprint(e))
+		} else {
+			logrus.Errorf("Caught FATAL error: %s", v)
+			debug.PrintStack()
```
Comment (cjellick) [inline api/response.go]: This is on purpose to get a pretty printed stack. If you use the logger, you get `\n` characters and no line breaks.

**Comment by yasker** [comment]
Overall looks good to me except the one I commented. Can merge it soon.

**Comment by cjellick** [comment]
@yasker Your comment didn't go to a specific line, so I don't have context for it.

**Comment by cjellick** [comment]
Agreed to put a TODO on the longhorn mountpoint issue and otherwise, @yasker said lgtm

**Pull request merged**